mqtt_as/mqtt_as.py

```diff
@@ -390,22 +396,17 @@ def close(self):
     async def subscribe(self, topic, qos):
         pkt = bytearray(b"\x82\0\0\0")
         pid = newpid(self.pid)
-        self.sub_pids.add(pid)  # PID's awaiting suback
+        self.pid = pid  # will otherwise result in multiple operations having the same pid
+        self.rcv_pids.add(pid)
         self.dprint("subscribe", topic, "pid", pid)
         struct.pack_into("!BH", pkt, 1, 2 + 2 + len(topic) + 1, pid)
-        self.pkt = pkt
         async with self.lock:
             await self._as_write(pkt)
             await self._send_str(topic)
             await self._as_write(qos.to_bytes(1, "little"))
 
-        t = ticks_ms()
-        while pid in self.sub_pids:  # Wait for SUBACK
-            await asyncio.sleep_ms(200)
-            if self._timeout(t):
-                self.dprint("subscribe timeout", topic)
-                raise OSError(-1)
-        self.dprint("subscribe done", topic)
+        if not await self._await_pid(pid):
```

Comment:
By using _await_pid the mechanism got extended to "if self._timeout(t) and not self.isconnected()" instead of previous line 405. A subscribe should terminate immediately if the connection breaks, just like the publish.